Mathlib/Algebra/Group/UniqueProds.lean

```diff
@@ -375,3 +387,150 @@ instance {ι} (G : ι → Type*) [∀ i, AddZeroClass (G i)] [∀ i, UniqueSums
+    obtain ⟨p1, h1, p2, h2, hne, hi1, hi2⟩ := uniqueMul_of_one_lt_card (Nat.one_lt_mul_iff.mpr
+      ⟨card_pos.2 (hA.image _), card_pos.2 (hB.image _), hc.imp And.left And.left⟩)
+    simp_rw [mem_product, mem_image, ← filter_nonempty_iff] at h1 h2
+    replace h1 := UniqueMul_of_TwoUniqueMul ?_ h1.1 h1.2
+    replace h2 := UniqueMul_of_TwoUniqueMul ?_ h2.1 h2.2
+    · obtain ⟨a1, ha1, b1, hb1, hu1⟩ := h1
+      obtain ⟨a2, ha2, b2, hb2, hu2⟩ := h2
+      rw [mem_filter] at ha1 hb1 ha2 hb2
+      simp_rw [mem_product]
+      refine ⟨(a1, b1), ⟨ha1.1, hb1.1⟩, (a2, b2), ⟨ha2.1, hb2.1⟩, ?_,
+        UniqueMul.of_image_filter (Pi.evalMulHom G i) ha1.2 hb1.2 hi1 hu1,
+        UniqueMul.of_image_filter (Pi.evalMulHom G i) ha2.2 hb2.2 hi2 hu2⟩
+      contrapose! hne; rw [Prod.mk.inj_iff] at hne ⊢
+      rw [← ha1.2, ← hb1.2, ← ha2.2, ← hb2.2, hne.1, hne.2]; exact ⟨rfl, rfl⟩
+    all_goals
+      rcases hc with hc | hc; · exact ihA _ (hc.2 _)
+      by_cases hA : A.filter (· i = _) = A; rw [hA]
+      exacts [ihB _ (hc.2 _), ihA _ ((A.filter_subset _).ssubset_of_ne hA)]
+
+open ULift in
+instance [TwoUniqueProds G] [TwoUniqueProds H] : TwoUniqueProds (G × H) := by
+  have : ∀ b, TwoUniqueProds (I G H b) := Bool.rec ?_ ?_
+  · exact mulHom_image_of_injective (downMulHom H) down_injective ‹_›
+  · refine mulHom_image_of_injective (Prod.upMulHom G H) (fun x y he => Prod.ext ?_ ?_)
+      (instTwoUniqueProdsForAllInstMul <| I G H) <;> apply up_injective
+    exacts [congr_fun he false, congr_fun he true]
+  · exact mulHom_image_of_injective (downMulHom G) down_injective ‹_›
+
+open Finset MulOpposite in
+@[to_additive]
+theorem of_mulOpposite (h : TwoUniqueProds Gᵐᵒᵖ) : TwoUniqueProds G where
+  uniqueMul_of_one_lt_card hc := by
+    let f : G ↪ Gᵐᵒᵖ := ⟨op, op_injective⟩
+    rw [← card_map f, ← card_map f, mul_comm] at hc
+    obtain ⟨p1, h1, p2, h2, hne, hu1, hu2⟩ := h.uniqueMul_of_one_lt_card hc
+    simp_rw [mem_product] at h1 h2 ⊢
+    refine ⟨(_, _), ⟨?_, ?_⟩, (_, _), ⟨?_, ?_⟩, ?_, hu1.of_mulOpposite, hu2.of_mulOpposite⟩
+    pick_goal 5
+    · contrapose! hne; rw [Prod.ext_iff] at hne ⊢
+      exact ⟨unop_injective hne.2, unop_injective hne.1⟩
+    all_goals apply (mem_map' f).mp
+    exacts [h1.2, h1.1, h2.2, h2.1]
+
+@[to_additive] instance [h : TwoUniqueProds G] : TwoUniqueProds Gᵐᵒᵖ :=
+  of_mulOpposite <| (mulHom_image_iff <| MulEquiv.opOp G).mp h
+
+open Finset in
+-- see Note [lower instance priority]
+/-- This instance asserts that if `G` has a right-cancellative multiplication, a linear order, and
+  multiplication is strictly monotone w.r.t. the second argument, then `G` has `TwoUniqueProds`. -/
+@[to_additive
+  "This instance asserts that if `G` has a right-cancellative addition, a linear order,
+  and addition is strictly monotone w.r.t. the second argument, then `G` has `TwoUniqueSums`." ]
+instance (priority := 100) of_Covariant_right [IsRightCancelMul G]
```

Comment:
I was expecting to see `Finset.min'_lt_max'_of_card` somewhere in this proof: does it not help?

Reply:
Indeed, [the proof using `min'_lt_max'`](https://github.com/leanprover-community/mathlib4/pull/7169/commits/5e10594daffcae239d69b2a3fef847a2ddc21277#diff-ebdf20b185bac5adc5a22ffdf53e7dfa74bb23fb34dfa47a761bb9dfa66c9fc1L494-R501) is shorter and no longer depends on `UniqueMul.iff_card_le_one` whose proof takes a few more lines. Thanks! We still need to use the existence of one unique pair to deduce `1 < (A*B).card` from `1 < (A×ˢB).card` though.